qt/scientific_interfaces/Indirect/test/IndirectDataTablePresenterTest.h

```diff
@@ -0,0 +1,368 @@
+  test_that_the_cellChanged_signal_will_set_the_models_excludeRegion_when_the_relevant_column_is_changed() {
+    int const excludeRegionColumn(4);
+
+    EXPECT_CALL(*m_model, setExcludeRegion("0-4", 0, 0)).Times(1);
+
+    m_table->item(0, excludeRegionColumn)->setText("0-4");
+  }
+
+  void
+  test_that_the_cellChanged_signal_will_set_the_models_startX_in_every_row_when_the_relevant_column_is_changed() {
+    int const startXColumn(2);
+
+    m_table->item(0, startXColumn)->setText("1.5");
+
+    for (auto row = 0; row < m_table->rowCount(); ++row)
+      TS_ASSERT_EQUALS(m_table->item(row, startXColumn)->text().toStdString(),
+                       "1.5")
+  }
+
+  void
+  test_that_the_cellChanged_signal_will_set_the_models_endX_in_every_row_when_the_relevant_column_is_changed() {
+    int const endXColumn(3);
+
+    m_table->item(0, endXColumn)->setText("2.5");
+
+    for (auto row = 0; row < m_table->rowCount(); ++row)
+      TS_ASSERT_EQUALS(m_table->item(row, endXColumn)->text().toStdString(),
+                       "2.5")
+  }
+
+  void
+  test_that_the_cellChanged_signal_will_set_the_models_excludeRegion_in_every_row_when_the_relevant_column_is_changed() {
+    int const excludeRegionColumn(4);
+
+    m_table->item(0, excludeRegionColumn)->setText("2-4");
+
+    for (auto row = 0; row < m_table->rowCount(); ++row)
+      TS_ASSERT_EQUALS(
+          m_table->item(row, excludeRegionColumn)->text().toStdString(), "2-4")
+  }
+
+  ///----------------------------------------------------------------------
+  /// Unit Tests that test the methods and slots of the presenter
+  ///----------------------------------------------------------------------
+
+  void
+  test_that_tableDatasetsMatchModel_returns_false_if_the_number_of_data_positions_is_not_equal_to_the_numberOfWorkspaces() {
+    std::size_t const numberOfWorkspaces(2);
+    ON_CALL(*m_model, numberOfWorkspaces())
+        .WillByDefault(Return(numberOfWorkspaces));
+
+    EXPECT_CALL(*m_model, numberOfWorkspaces())
+        .Times(1)
+        .WillOnce(Return(numberOfWorkspaces));
+
+    TS_ASSERT(!m_presenter->tableDatasetsMatchModel());
+  }
+
+  void
+  test_that_tableDatasetsMatchModel_returns_true_if_the_table_datasets_match_the_model() {
+    EXPECT_CALL(*m_model, numberOfWorkspaces()).Times(1).WillOnce(Return(0));
+    TS_ASSERT(m_presenter->tableDatasetsMatchModel());
+  }
+
+  void
+  test_that_addData_will_add_new_data_if_the_index_is_smaller_than_the_number_of_data_positions() {
+    std::size_t const index(0);
+
+    ON_CALL(*m_model, numberOfWorkspaces()).WillByDefault(Return(2));
+
+    EXPECT_CALL(*m_model, numberOfWorkspaces()).Times(1);
+
+    ExpectationSet getRanges =
+        EXPECT_CALL(*m_model, getFittingRange(index, 0)).Times(1);
+    for (auto spectrum = 1; spectrum < m_table->rowCount(); ++spectrum)
+      getRanges += EXPECT_CALL(*m_model, getFittingRange(index, spectrum))
+                       .Times(1)
+                       .After(getRanges);
```

Comment:
What is .After?

Reply:
It means it is expecting `getFittingRange` to be called after the expectation assigned to `getRanges`. In this case it is expecting getFittingRange(0, 1) to be called after getFittingRange(0, 0). Then it checks that getFittingRange(0,2) is called after getFittingRange(0, 1) and getFittingRange(0, 0), and so on.